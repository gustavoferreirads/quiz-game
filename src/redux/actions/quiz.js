import axios from 'axios';
import { createAction } from 'redux-actions';
import { ANSWER, QUESTIONS } from './index';

export function doAnswer(answer) {
  return dispatch => {
    dispatch({
      type: ANSWER,
      answer,
    });
  };
}

function get(amount = 10, difficulty = 'hard', type = 'boolean') {
  return axios.get('api.php', {
    params: {
      amount,
      difficulty,
      type,
    },
  });
}
export const fetchQuestions = createAction(QUESTIONS, get);
