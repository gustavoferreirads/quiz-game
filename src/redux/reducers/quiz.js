import { AllHtmlEntities } from 'html-entities';
import { ANSWER, QUESTIONS } from '../actions';

const initialState = {
  questions: [],
  size: 0,
  current: { index: -1, wrapQuestion: {} },
  answers: [],
  score: 0,
  percent: 0,
};

const generateResult = state => {
  const { questions } = state;
  let score = 0;
  const answers = questions.map((question, index) => {
    const isCorrect = state.answers[index] === question.correct_answer;
    score = isCorrect ? score + 1 : score;
    return { ...question, isCorrect };
  });
  const percent = (score / answers.length) * 100;
  return { ...initialState, score, answers, percent };
};

const nextQuestion = state => {
  const { current, questions } = state;

  const index = current.index + 1;

  if (index < questions.length) {
    const wrapQuestion = questions[index];

    return {
      ...state,
      current: { index, wrapQuestion, position: index + 1 },
    };
  }

  return generateResult(state);
};

const reducer = (state = initialState, action) => {
  if (action.type === QUESTIONS) {
    const { payload } = action;
    const questions = payload.data.results.map(q => ({
      ...q,
      question: new AllHtmlEntities().decode(q.question),
    }));
    const newState = { ...initialState, questions, size: questions.length };
    return nextQuestion(newState);
  }

  if (action.type === ANSWER) {
    const { answers } = state;

    answers.push(action.answer);

    return nextQuestion({
      ...state,
      answers,
    });
  }

  return state;
};

export default reducer;
