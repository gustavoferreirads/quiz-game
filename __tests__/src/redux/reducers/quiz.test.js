import quiz from '../../../../src/redux/reducers/quiz';
import initialState from './initial';
import questionsAction from './questions';
import answerAction from './answer';

it('should find questions and define current question', () => {
  expect(quiz(initialState, questionsAction)).toMatchSnapshot();
});

it('should do answer', () => {
  const state = quiz(initialState, questionsAction);
  expect(quiz(state, answerAction)).toMatchSnapshot();
});

it('should do final answer and generate result', () => {
  let state = quiz(initialState, questionsAction);
  state = quiz(state, answerAction);
  expect(quiz(state, answerAction)).toMatchSnapshot();
});
