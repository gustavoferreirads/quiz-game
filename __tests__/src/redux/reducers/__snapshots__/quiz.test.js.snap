// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should do answer 1`] = `
Object {
  "answers": Array [
    "True",
  ],
  "current": Object {
    "index": 1,
    "position": 2,
    "wrapQuestion": Object {
      "category": "General Knowledge",
      "correct_answer": "True",
      "difficulty": "hard",
      "incorrect_answers": Array [
        "False",
      ],
      "question": "In Scandinavian languages, the letter Å means river.",
      "type": "boolean",
    },
  },
  "percent": 0,
  "questions": Array [
    Object {
      "category": "Entertainment: Film",
      "correct_answer": "False",
      "difficulty": "hard",
      "incorrect_answers": Array [
        "True",
      ],
      "question": "\\"Cube\\", \\"Cube 2: Hypercube\\" and \\"Cube Zero\\" were directed by the same person.",
      "type": "boolean",
    },
    Object {
      "category": "General Knowledge",
      "correct_answer": "True",
      "difficulty": "hard",
      "incorrect_answers": Array [
        "False",
      ],
      "question": "In Scandinavian languages, the letter Å means river.",
      "type": "boolean",
    },
  ],
  "score": 0,
  "size": 2,
}
`;

exports[`should do final answer and generate result 1`] = `
Object {
  "answers": Array [
    Object {
      "category": "Entertainment: Film",
      "correct_answer": "False",
      "difficulty": "hard",
      "incorrect_answers": Array [
        "True",
      ],
      "isCorrect": false,
      "question": "\\"Cube\\", \\"Cube 2: Hypercube\\" and \\"Cube Zero\\" were directed by the same person.",
      "type": "boolean",
    },
    Object {
      "category": "General Knowledge",
      "correct_answer": "True",
      "difficulty": "hard",
      "incorrect_answers": Array [
        "False",
      ],
      "isCorrect": true,
      "question": "In Scandinavian languages, the letter Å means river.",
      "type": "boolean",
    },
  ],
  "current": Object {
    "index": -1,
    "wrapQuestion": Object {},
  },
  "percent": 50,
  "questions": Array [],
  "score": 1,
  "size": 0,
}
`;

exports[`should find questions and define current question 1`] = `
Object {
  "answers": Array [],
  "current": Object {
    "index": 0,
    "position": 1,
    "wrapQuestion": Object {
      "category": "Entertainment: Film",
      "correct_answer": "False",
      "difficulty": "hard",
      "incorrect_answers": Array [
        "True",
      ],
      "question": "\\"Cube\\", \\"Cube 2: Hypercube\\" and \\"Cube Zero\\" were directed by the same person.",
      "type": "boolean",
    },
  },
  "percent": 0,
  "questions": Array [
    Object {
      "category": "Entertainment: Film",
      "correct_answer": "False",
      "difficulty": "hard",
      "incorrect_answers": Array [
        "True",
      ],
      "question": "\\"Cube\\", \\"Cube 2: Hypercube\\" and \\"Cube Zero\\" were directed by the same person.",
      "type": "boolean",
    },
    Object {
      "category": "General Knowledge",
      "correct_answer": "True",
      "difficulty": "hard",
      "incorrect_answers": Array [
        "False",
      ],
      "question": "In Scandinavian languages, the letter Å means river.",
      "type": "boolean",
    },
  ],
  "score": 0,
  "size": 2,
}
`;
